Sources/Rendering/Core/Picker/index.js

```diff
@@ -89,9 +89,171 @@ function vtkPicker(publicAPI, model) {
+      return;
+    }
+
+    //  Loop over all props.  Transform ray (defined from position of
+    //  camera to selection point) into coordinates of mapper (not
+    //  transformed to actors coordinates!  Reduces overall computation!!!).
+    //  Note that only vtkProp3D's can be picked by vtkPicker.
+    const props = model.pickFromList ? model.pickList : renderer.getActors();
+
+    const scale = [];
+    props.forEach((prop) => {
+      const mapper = prop.getMapper();
+      pickable =
+        prop.getNestedPickable() &&
+        prop.getNestedVisibility() &&
+        prop.getProperty().getOpacity() > 0.0;
+
+      //  If actor can be picked, get its composite matrix, invert it, and
+      //  use the inverted matrix to transform the ray points into mapper
+      //  coordinates.
+      if (pickable) {
+        model.transformMatrix = prop.getMatrix().slice(0);
+
+        mat4.transpose(model.transformMatrix, model.transformMatrix);
+        mat4.invert(model.transformMatrix, model.transformMatrix);
+
+        // Extract scale
+        const col1 = [
+          model.transformMatrix[0],
+          model.transformMatrix[1],
+          model.transformMatrix[2],
+        ];
+        const col2 = [
+          model.transformMatrix[4],
+          model.transformMatrix[5],
+          model.transformMatrix[6],
+        ];
+        const col3 = [
+          model.transformMatrix[8],
+          model.transformMatrix[9],
+          model.transformMatrix[10],
+        ];
+
+        scale[0] = vtkMath.norm(col1);
+        scale[1] = vtkMath.norm(col2);
+        scale[2] = vtkMath.norm(col3);
+
+        vec4.transformMat4(p1Mapper, p1World, model.transformMatrix);
+        vec4.transformMat4(p2Mapper, p2World, model.transformMatrix);
+
+        p1Mapper[0] /= p1Mapper[3];
+        p1Mapper[1] /= p1Mapper[3];
+        p1Mapper[2] /= p1Mapper[3];
+
+        p2Mapper[0] /= p2Mapper[3];
+        p2Mapper[1] /= p2Mapper[3];
+        p2Mapper[2] /= p2Mapper[3];
+
+        for (let i = 0; i < 3; i++) {
+          ray[i] = p2Mapper[i] - p1Mapper[i];
+        }
+
+        //  Have the ray endpoints in mapper space, now need to compare this
+        //  with the mapper bounds to see whether intersection is possible.
+        //
+        //  Get the bounding box of the modeller.  Note that the tolerance is
+        //  added to the bounding box to make sure things on the edge of the
+        //  bounding box are picked correctly.
+        const bounds = mapper
+          ? vtkBoundingBox.inflate(mapper.getBounds(), tolerance)
+          : [0, 0, 0, 0, 0, 0];
+
+        if (
+          vtkBoundingBox.intersectBox(bounds, p1Mapper, ray, hitPosition, t)
+        ) {
+          t[0] = publicAPI.intersectWithLine(
+            p1Mapper,
+            p2Mapper,
+            tolerance * 0.333 * (scale[0] + scale[1] + scale[2]),
+            prop,
+            mapper
+          );
+
+          if (t[0] < Number.MAX_VALUE) {
+            const p = [];
```

Comment:
`scale` is predefined before the for loop. Why not do the same for `p` ?

Reply:
I assume some variables are defined outside of the closure to avoid allocating/destroying every time the function is called by `forEach`?
Is this really a big game changer for performance? What's the policy/best practices that should be applied regarding this in VTK.js?
Naively I would say either we define everything in the closure, either everything outside.

Reply:
There is no rule in VTK.js.
I think it should be declared outside a `for` loop in `time sensitive` functions.
And I agree with you that it should be all or none. Not a mix of it.
Because the picking is `time sensitive` I therefore suggest to do pre-allocation everywhere.

Reply:
is the `for...of` syntax allowed in VTK.js?